File: 76. Minimum Window Substring.py
from collections import Counter, defaultdict
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not s or not t:
            return ''
        
        dict_t = Counter(t)
        required = len(dict_t)
        
        filtered_s = []
        for i, char in enumerate(s):
            if char in dict_t:
                filtered_s.append((char, i))
        
        l,r = 0,0
        formed = 0
        window_count = defaultdict(int)
        
        ans = 2**31-1, None, None
        
        while r < len(filtered_s):
            char = filtered_s[r][0]
            window_count[char] += 1
            
            if window_count[char] == dict_t[char]:
                formed += 1
                
            while l <= r and formed == required:
                ch = filtered_s[l][0]
                
                end = filtered_s[r][1]
                start = filtered_s[l][1]
                if end - start + 1 < ans[0]:
                    ans = (end - start + 1, start, end)
                    
                window_count[ch] -= 1
                if window_count[ch] < dict_t[ch]:
                    formed -= 1
                l += 1
            r += 1
        return '' if ans[0] == 2**31-1 else s[ans[1]:ans[2]+1]

File: 76. Minimum Window Substring.java
import javafx.util.*;
class Solution {
    public String minWindow(String s, String t) {
        if (s.length()==0 || t.length()==0) {
            return "";
        }
        List<Pair<Integer, Character>> map = new ArrayList<>();
        HashMap<Character, Integer> tmap = new HashMap<>();
        for (char c : t.toCharArray()) {
            tmap.put(c, tmap.getOrDefault(c, 0)+1);
        }
        for (int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (tmap.containsKey(c)) {
                map.add(new Pair<>(i, c));
            }
        }
        int st=-1, ed=s.length();
        int l=0, r=0;
        int required = tmap.size(), formed=0;
        HashMap<Character, Integer> count = new HashMap<>();
        while (r < map.size()) {
            char c = map.get(r).getValue();
            count.put(c, count.getOrDefault(c, 0)+1);
            if (tmap.containsKey(c) && count.get(c).intValue() == tmap.get(c).intValue()) {
                formed++;
            }
            while (l <= r && formed==required) {
                c = map.get(l).getValue();
                int end = map.get(r).getKey();
                int start = map.get(l).getKey();
                if (end-start+1 < ed-st+1) {
                    ed = end;
                    st = start;
                }
                
                count.put(c, count.get(c)-1);
                if (tmap.containsKey(c) && count.get(c).intValue() < tmap.get(c).intValue()) {
                    formed--;
                }
                l++;
            }
            r++;
        }
        return ed-st > s.length()? "" : s.substring(st, ed+1);
    }
}